### ISSUE TITLE
Aggregating data with multiple fields

### ISSUE THREAD

**Issue by bbczeuz** [influxdata/kapacitor#696]
I'm trying to aggregate all fields of a certain series to time steps of 1h. With continuous queries, this worked as follows:
```
CREATE CONTINUOUS QUERY cq_1h_net_sec ON telegraf
BEGIN 
  SELECT
    mean(bytes_sec_recv) as bytes_sec_recv_avg, 
    max(bytes_sec_recv) as bytes_sec_recv_max, 
    min(bytes_sec_recv) as bytes_sec_recv_min, 
    percentile(bytes_sec_recv, 95) as bytes_sec_recv_perc95,  
    mean(bytes_sec_sent) as bytes_sec_sent_avg, 
    max(bytes_sec_sent) as bytes_sec_sent_max, 
    min(bytes_sec_sent) as bytes_sec_sent_min
    percentile(bytes_sec_sent, 95) as bytes_sec_sent_perc95, 
  INTO "store1y"."net_sec"
  FROM "store7d".net_sec 
  GROUP BY *, time(1h)
END
```

In the example, I process two fields, but I'm searching for a more general solution that processes all fields, a regex match or the like. With kapacitor, I don't see a solution yet; maybe splitting, processing and joining fields?
```
var val_raw = stream
    |from()
        .database('telegraf')
        .retentionPolicy('store7d')
        .measurement('net_sec')
        .where(lambda: "interface" == 'ens192')
        .groupBy(*)
    |default()
        .field('bytes_sec_recv', 0)
        .field('bytes_sec_sent', 0)

var windowed_vals = val_raw
    |window()
        .period(30s)
        .every(30s)
        .align()

var bytes_sec_recv_avg = windowed_vals
    |mean('bytes_sec_recv')
        .as('bytes_sec_recv_avg')

var bytes_sec_sent_avg = windowed_vals
    |mean('bytes_sec_sent')
        .as('bytes_sec_sent_avg')

bytes_sec_recv_avg
    |union(bytes_sec_sent_avg)
    |influxDBOut()
        .database('telegraf')
        .retentionPolicy('store1y')
        .measurement('net_sec')
        .precision('s')
```

Is this the way to go? I'm not sure if this scales well, as there are ~8 fields (rx and tx of bytes, packets, errors and drops) and 6 aggregation functions (min, max, avg, percentile (25, 75, 95)) per telegraf net series, leading to 48 variables...
Something like 
```
    |mean('bytes_sec_sent', 'bytes_sec_recv')
        .as('bytes_sec_sent_avg', 'bytes_sec_recv_avg')
```
or
```
    |mean('*')
        .postfix('_avg')
```
would be cool.

**Comment by nathanielc** [comment]
>  With kapacitor, I don't see a solution yet; maybe splitting, processing and joining fields?

Yes, this is currently how it can be done in Kapacitor. Like you say it doesn't scale well since it requires lots of coding and internally copying of data. 

I like your suggested solutions.

Also you will want to use join instead of union of you want all the fields to be on the same point, (unless they have the same timestamp and InfluxDB will merge them internally)

**Comment by phemmer** [comment]
I too am highly interested in using multiple aggregates on the same window. However the proposed syntax wouldn't suit my use case as we're looking to use multiple different aggregate functions (`mean`, `sum`, `max`, etc). However I'm not getting any bright ideas on a syntax that would work.

The current `join` method isn't too terrible in concept, except that it forces you to have a prefix. It's very cumbersome to have to specify a prefix for each aggregate when they all came from the same measurement & window.
E.G.: `bytes_sec_recv_avg|join(bytes_sec_sent_avg).as('bytes_sec_recv_avg','bytes_sec_sent_avg')` and then referencing it as `"bytes_sec_sent_avg.bytes_sec_sent_avg"` is just icky, and meaningless join names such as `j1`, `j2`, etc, aren't much better.

**Comment by nathanielc** [comment]
@phemmer It might be helpful to point out that with a recent version (don't remember off the top of my head, which one) the join node now allows for more control over the prefix names and delimiter.

Also note that you can name the field something generic as well like `value`. Then the code reads like,

```
bytes_sec_recv_avg
    |join(bytes_sec_sent_avg)
        .as('bytes_sec_recv_avg','bytes_sec_sent_avg')
    |eval(lambda: "bytes_sec_recv_avg.value" + "bytes_sec_sent_avg.value")
    |...
```

This is how I have been doing it. I would like a cleaner method as well but I haven't really figured out a better syntax, yet. One possible improvement would be to make the `.as` property optional, (which has been suggested before, #795), and then its the user's responsibility to make sure the field names are unique across the joined nodes.

The current syntax while verbose can handle every use case I have seen so far expect where the field names themselves are dynamic. Dynamic fields is not common and can be handled with a UDF currently so for now I am inclined to leave things as they are.

**Comment by Yaski** [comment]
@nathanielc 
Is it possible to do many aggregations on the same window? For example:
```
var data = windowed_vals
    |mean('bytes_sec_recv')
        .as('bytes_sec_recv_avg')
    |mean('bytes_sec_sent')
        .as('bytes_sec_sent_avg')

```
So data should have on point with two fields - "bytes_sec_recv" and "bytes_sec_sent_avg"

Joining is very time consuming even if I have only one point in each source and both of them have occurred at the same time.
I'm looking for a simpler method that I can use in my tick script - I want to compare fields from two points from two different sources. It is very easy to achieve in any programming language but in kapacitor it looks like I should merge data before use.

**Comment by yoitsbenco** [comment]
I believe I have this same/similar problem where I have grouped data points by host (required because I calculate derivative by each host), but seemingly no way of dynamically merging (i.e., *without* querying each host individually and then doing a hardcoded `hostA.join(hostB, hostC, hostD, ... hostZ`).

Is there really no way to merge grouped data without using the `join`?
